src/v/kafka/server/protocol.cc

```diff
@@ -95,10 +96,35 @@ coordinator_ntp_mapper& protocol::coordinator_mapper() {
 }
 
 config::broker_authn_method get_authn_method(const net::connection& conn) {
+    // Global enable_sasl is deprecated, replaced by broker_authn_method == sasl
+    // If authn_method is set on the endpoint
+    //    Use it
+    // Else if enable_authorizer is not set
+    //    Use the old behaviour - sasl if enable_sasl
+    // Else if has mtls mapping rules
+    //    Use mtls_identity
+    // Else
+    //    Disable AuthN
+
+    std::optional<config::broker_authn_method> authn_method;
+    auto n = conn.name();
+    const auto& kafka_api = config::node().kafka_api.value();
+    auto ep_it = std::find_if(
+      kafka_api.begin(),
+      kafka_api.end(),
+      [&n](const config::broker_authn_endpoint& ep) { return ep.name == n; });
+    if (ep_it != kafka_api.end()) {
+        authn_method = ep_it->authn_method;
+    }
+    if (authn_method.has_value()) {
+        return *authn_method;
+    }
     const auto& config = config::shard_local_cfg();
-    if (config.enable_sasl()) {
+    // if enable_authorization is set, ignore enable_sasl
+    if (!config.enable_authorization().has_value() && config.enable_sasl()) {
         return config::broker_authn_method::sasl;
     }
+    // mtls_identity is currently predicated on having mapping rules
     if (conn.get_principal_mapping().has_value()) {
         return config::broker_authn_method::mtls_identity;
     }
```

Comment:
The last default path of returning ::none is a bit strange: if the user has set enable_authorization but not set a principal mapping, we disable authentication?  Maybe this is the path that should set it to SASL?  Otherwise we end up in a state where authorization is required but authentication is off :confused:

Reply:
This is an intermediate step to keep the changes compiling.

To enable `sasl`, do the same configuration as before, i.e., `enable_sasl: true`, and either `enable_authorization: nullopt` (effectively as before) or by setting `authentication_method: sasl`.

Reply:
EDIT: does John's comment still apply after the changes later?